pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,917 @@
+		if name == shareName {
+			return instanceInfo, nil
+		}
+	}
+	klog.Infof("Appending share %q to %q instanceInfo's assigned share list", shareName, instanceInfoClone.Name)
+	instanceInfoClone.Status.ShareNames = append(instanceInfoClone.Status.ShareNames, shareName)
+
+	return recon.updateInstanceInfoStatus(context.TODO(), instanceInfoClone)
+}
+
+func (recon *multishareReconciler) assignInstanceToShareInfo(shareInfo *v1alpha1.ShareInfo, instanceURI string) (*v1alpha1.ShareInfo, error) {
+	shareInfoClone := shareInfo.DeepCopy()
+	if shareInfoClone.Status == nil {
+		shareInfoClone.Status = &v1alpha1.ShareInfoStatus{}
+	}
+	shareInfoClone.Status.InstanceHandle = instanceURI
+	klog.Infof("Try to assign share %q to instance %q", shareInfo.Name, instanceURI)
+	return recon.updateShareInfoStatus(context.TODO(), shareInfoClone)
+}
+
+// createAndUpdateInstanceInfos create instanceInfo objects if needed and updates their statuses to match with reality.
+// InstanceInfo objects in the returned map must be treated as read only.
+func (recon *multishareReconciler) createAndUpdateInstanceInfos(instances []*file.MultishareInstance, instanceShares map[string][]*file.Share) map[string]*v1alpha1.InstanceInfo {
+	instanceInfoMap := make(map[string]*v1alpha1.InstanceInfo)
+
+	for _, instance := range instances {
+		instanceURI, err := file.GenerateMultishareInstanceURI(instance)
+		if err != nil {
+			klog.Errorf("Couldn't generate instanceURI: %v for instance %q", err, instance.Name)
+			continue
+		}
+		iiName := util.InstanceURIToInstanceInfoName(instanceURI)
+
+		instanceInfo, err := recon.instanceLister.Get(iiName)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				klog.Errorf("Error getting instanceInfo %q from informer: %v", iiName, err)
+				continue
+			}
+			instanceInfo, err = recon.clientset.MultishareV1alpha1().InstanceInfos().Get(context.TODO(), iiName, metav1.GetOptions{})
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					klog.Errorf("Error getting instanceInfo %q from api server: %v", iiName, err)
+					continue
+				}
+				klog.V(4).Infof("InstanceInfo object for instance %q not found in API server", instance.Name)
+				instanceInfo = nil
+			}
+			klog.V(4).Infof("InstanceInfo object for instance %q not found in informer cache but found in API server", instance.Name)
+		}
+
+		instanceInfo, err = recon.reconstructInstanceInfo(iiName, instance, instanceInfo)
+		if err != nil {
+			klog.Errorf("Error creating InstanceInfo %q: %v", iiName, err)
+			continue
+		}
+		instanceInfoMap[instanceURI] = instanceInfo
+
+		instanceInfo, err = recon.maybeUpdateInstanceInfoStatus(instance, instanceInfo, instanceShares)
+		if err != nil {
+			klog.Errorf("Error updating instanceInfo %q: %v", iiName, err)
+			continue
+		}
+		instanceInfoMap[instanceURI] = instanceInfo
+	}
+
+	return instanceInfoMap
+}
+
+// createAndUpdateShareInfos create shareInfo objects if needed and updates their statuses to match with reality.
```

Comment:
nit: ... to match with the real state of the world

Reply:
done